File: frontend/src/types/geo.ts
export interface CheckResult {
  category: string;
  status: string; // PASS, WARN, FAIL, INFO
  message: string;
  fix?: string;
  /** i18n key path under result.checks.* when the check was emitted
   *  via geo_checker.emit_check(). Absent for un-migrated legacy checks. */
  message_key?: string | null;
  /** Interpolation params for message_key (e.g. {count: 12, url: "..."}). */
  message_params?: Record<string, unknown> | null;
  /** i18n key under result.fixes.* for the fix recommendation. Absent for
   *  legacy fix() calls that haven't been migrated to emit_fix(). */
  fix_key?: string | null;
  /** Interpolation params for fix_key. */
  fix_params?: Record<string, unknown> | null;
}

export interface GeoTestRequest {
  url: string;
  include_fix: boolean;
}

export interface GeoTestResult {
  url: string;
  score: number;
  grade: string;
  checks: CheckResult[];
  summary: {
    pass_count: number;
    warn_count: number;
    fail_count: number;
    info_count: number;
    total_checks: number;
  };
  /** Membership slug under which this check ran ('free' | 'pro' | 'starter' | 'growth' | 'scale'). */
  tier?: string | null;
  /** Category labels that were NOT executed for this tier. Empty for unlocked tiers. */
  locked_categories?: string[] | null;
}

File: backend/geo/models/geo.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any

class GeoTestRequest(BaseModel):
    """Request model for GEO test"""
    url: str = Field(..., description="Website URL to test")
    include_fix: bool = Field(default=True, description="Include fix recommendations")
    force_refresh: bool = Field(
        default=False,
        description=(
            "Skip L1 Redis + L2 DB cache and run fresh. Used by the 'verify "
            "after fix' flow on the result page so users see updated scores "
            "immediately instead of the 24h-stale cached report."
        ),
    )

class CheckResult(BaseModel):
    """Result of a single check"""
    category: str
    status: str  # PASS, WARN, FAIL, INFO
    message: str
    fix: Optional[str] = None
    # i18n key + params emitted by geo_checker.emit_check(). When present,
    # the frontend renders t(message_key, message_params); when absent
    # (legacy un-migrated checks), the frontend falls back to `message`.
    message_key: Optional[str] = None
    message_params: Optional[Dict[str, Any]] = None
    # i18n key + params for the fix recommendation, emitted by emit_fix().
    # Same fallback: when absent the frontend renders plain `fix` text.
    fix_key: Optional[str] = None
    fix_params: Optional[Dict[str, Any]] = None

class GeoTestResult(BaseModel):
    """Response model for GEO test result"""
    url: str
    score: int
    grade: str
    checks: List[CheckResult]
    summary: Dict[str, Any]
    tier: Optional[str] = None  # 'free' | 'pro' | 'starter' | 'growth' | 'scale'
    locked_categories: Optional[List[str]] = None  # categories hidden from this tier
